{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob\n",
    "import os\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#read in csv file\n",
    "sample_sheet = pd.read_csv('sample_sheet.csv')\n",
    "#sample_sheet = pd.read_csv('sample_sheet_new.csv')\n",
    "ANNOTATIONS_DIR = \"/Volumes/bioinformatics/pipeline/annotation/pipeline/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['/Volumes/bioinformatics/pipeline/annotation/pipeline/Crassostrea_gigas_v1/Crassostrea_gigas_v1.gtf']"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "glob.glob(os.path.join(ANNOTATIONS_DIR,\"Crassostrea_gigas_v1\",\"*.gtf\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['.DS_Store',\n",
       " 'Mus_musculus.GRCm38.91.gtf.refflat.txt',\n",
       " 'Mus_musculus.GRCm38.dna.primary_assembly.chromap',\n",
       " 'Mus_musculus.GRCm38.dna.primary_assembly.fa',\n",
       " 'Mus_musculus.GRCm38_chromInfo.txt',\n",
       " 'Mus_musculus.GRCm38.91.refflat.txt',\n",
       " 'Mus_musculus.GRCm38.91.gtf',\n",
       " 'star',\n",
       " 'Mus_musculus.GRCm38.91.ribosomalIntervalsWheader.txt']"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.listdir(os.path.join(ANNOTATIONS_DIR,\"GRCm38.91\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['containers',\n",
       " 'ICACviruses',\n",
       " '10x',\n",
       " 'Crassostrea_virginica_3.0',\n",
       " '.DS_Store',\n",
       " 'GRCm38.91_tdTomato_ZsGreen',\n",
       " 'GRCm38.91_tdTomato',\n",
       " 'GRCm38.91',\n",
       " 'GRCh38.91',\n",
       " 'TyGRCh38',\n",
       " 'Crassostrea_gigas_v1',\n",
       " 'GRCh37',\n",
       " 'GRCm38.91_SS78_SS95',\n",
       " 'GRCh37_ucsc']"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.listdir(ANNOTATIONS_DIR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>project</th>\n",
       "      <th>flowcell</th>\n",
       "      <th>reference</th>\n",
       "      <th>assay_type</th>\n",
       "      <th>read_type</th>\n",
       "      <th>input_path</th>\n",
       "      <th>output_path</th>\n",
       "      <th>flowcellID</th>\n",
       "      <th>sample_flowcell</th>\n",
       "      <th>R1</th>\n",
       "      <th>R2</th>\n",
       "      <th>species</th>\n",
       "      <th>genome</th>\n",
       "      <th>save_to_db</th>\n",
       "      <th>output_files</th>\n",
       "      <th>genomics_samples</th>\n",
       "      <th>workflowbatchId</th>\n",
       "      <th>batchFile</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>lib105608</td>\n",
       "      <td>P505-1</td>\n",
       "      <td>241231_VH00126_375_AACN5FKHV</td>\n",
       "      <td>mouse</td>\n",
       "      <td>bulkRNAseq</td>\n",
       "      <td>pe</td>\n",
       "      <td>/nfs/bioinformatics/pipeline/Illumina/241231_V...</td>\n",
       "      <td>/nfs/bioinformatics/pipeline/Illumina/241231_V...</td>\n",
       "      <td>AACN5FKHV</td>\n",
       "      <td>lib105608_AACN5FKHV</td>\n",
       "      <td>['/nfs/bioinformatics/pipeline/Illumina/241231...</td>\n",
       "      <td>['/nfs/bioinformatics/pipeline/Illumina/241231...</td>\n",
       "      <td>Mus_musculus</td>\n",
       "      <td>GRCm38.91</td>\n",
       "      <td>True</td>\n",
       "      <td>['/nfs/bioinformatics/pipeline/Illumina/241231...</td>\n",
       "      <td>['/nfs/bioinformatics/pipeline/Illumina/241231...</td>\n",
       "      <td>bri_P505-1_2025-01-02</td>\n",
       "      <td>/nfs/bioinformatics/pipeline/snakemakeLogs/202...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>lib105609</td>\n",
       "      <td>P505-1</td>\n",
       "      <td>241231_VH00126_375_AACN5FKHV</td>\n",
       "      <td>mouse</td>\n",
       "      <td>bulkRNAseq</td>\n",
       "      <td>pe</td>\n",
       "      <td>/nfs/bioinformatics/pipeline/Illumina/241231_V...</td>\n",
       "      <td>/nfs/bioinformatics/pipeline/Illumina/241231_V...</td>\n",
       "      <td>AACN5FKHV</td>\n",
       "      <td>lib105609_AACN5FKHV</td>\n",
       "      <td>['/nfs/bioinformatics/pipeline/Illumina/241231...</td>\n",
       "      <td>['/nfs/bioinformatics/pipeline/Illumina/241231...</td>\n",
       "      <td>Mus_musculus</td>\n",
       "      <td>GRCm38.91</td>\n",
       "      <td>True</td>\n",
       "      <td>['/nfs/bioinformatics/pipeline/Illumina/241231...</td>\n",
       "      <td>['/nfs/bioinformatics/pipeline/Illumina/241231...</td>\n",
       "      <td>bri_P505-1_2025-01-02</td>\n",
       "      <td>/nfs/bioinformatics/pipeline/snakemakeLogs/202...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>lib105610</td>\n",
       "      <td>P505-1</td>\n",
       "      <td>241231_VH00126_375_AACN5FKHV</td>\n",
       "      <td>mouse</td>\n",
       "      <td>bulkRNAseq</td>\n",
       "      <td>pe</td>\n",
       "      <td>/nfs/bioinformatics/pipeline/Illumina/241231_V...</td>\n",
       "      <td>/nfs/bioinformatics/pipeline/Illumina/241231_V...</td>\n",
       "      <td>AACN5FKHV</td>\n",
       "      <td>lib105610_AACN5FKHV</td>\n",
       "      <td>['/nfs/bioinformatics/pipeline/Illumina/241231...</td>\n",
       "      <td>['/nfs/bioinformatics/pipeline/Illumina/241231...</td>\n",
       "      <td>Mus_musculus</td>\n",
       "      <td>GRCm38.91</td>\n",
       "      <td>True</td>\n",
       "      <td>['/nfs/bioinformatics/pipeline/Illumina/241231...</td>\n",
       "      <td>['/nfs/bioinformatics/pipeline/Illumina/241231...</td>\n",
       "      <td>bri_P505-1_2025-01-02</td>\n",
       "      <td>/nfs/bioinformatics/pipeline/snakemakeLogs/202...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>lib105611</td>\n",
       "      <td>P505-1</td>\n",
       "      <td>241231_VH00126_375_AACN5FKHV</td>\n",
       "      <td>mouse</td>\n",
       "      <td>bulkRNAseq</td>\n",
       "      <td>pe</td>\n",
       "      <td>/nfs/bioinformatics/pipeline/Illumina/241231_V...</td>\n",
       "      <td>/nfs/bioinformatics/pipeline/Illumina/241231_V...</td>\n",
       "      <td>AACN5FKHV</td>\n",
       "      <td>lib105611_AACN5FKHV</td>\n",
       "      <td>['/nfs/bioinformatics/pipeline/Illumina/241231...</td>\n",
       "      <td>['/nfs/bioinformatics/pipeline/Illumina/241231...</td>\n",
       "      <td>Mus_musculus</td>\n",
       "      <td>GRCm38.91</td>\n",
       "      <td>True</td>\n",
       "      <td>['/nfs/bioinformatics/pipeline/Illumina/241231...</td>\n",
       "      <td>['/nfs/bioinformatics/pipeline/Illumina/241231...</td>\n",
       "      <td>bri_P505-1_2025-01-02</td>\n",
       "      <td>/nfs/bioinformatics/pipeline/snakemakeLogs/202...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>lib105612</td>\n",
       "      <td>P505-1</td>\n",
       "      <td>241231_VH00126_375_AACN5FKHV</td>\n",
       "      <td>mouse</td>\n",
       "      <td>bulkRNAseq</td>\n",
       "      <td>pe</td>\n",
       "      <td>/nfs/bioinformatics/pipeline/Illumina/241231_V...</td>\n",
       "      <td>/nfs/bioinformatics/pipeline/Illumina/241231_V...</td>\n",
       "      <td>AACN5FKHV</td>\n",
       "      <td>lib105612_AACN5FKHV</td>\n",
       "      <td>['/nfs/bioinformatics/pipeline/Illumina/241231...</td>\n",
       "      <td>['/nfs/bioinformatics/pipeline/Illumina/241231...</td>\n",
       "      <td>Mus_musculus</td>\n",
       "      <td>GRCm38.91</td>\n",
       "      <td>True</td>\n",
       "      <td>['/nfs/bioinformatics/pipeline/Illumina/241231...</td>\n",
       "      <td>['/nfs/bioinformatics/pipeline/Illumina/241231...</td>\n",
       "      <td>bri_P505-1_2025-01-02</td>\n",
       "      <td>/nfs/bioinformatics/pipeline/snakemakeLogs/202...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      sample project                      flowcell reference  assay_type  \\\n",
       "0  lib105608  P505-1  241231_VH00126_375_AACN5FKHV     mouse  bulkRNAseq   \n",
       "1  lib105609  P505-1  241231_VH00126_375_AACN5FKHV     mouse  bulkRNAseq   \n",
       "2  lib105610  P505-1  241231_VH00126_375_AACN5FKHV     mouse  bulkRNAseq   \n",
       "3  lib105611  P505-1  241231_VH00126_375_AACN5FKHV     mouse  bulkRNAseq   \n",
       "4  lib105612  P505-1  241231_VH00126_375_AACN5FKHV     mouse  bulkRNAseq   \n",
       "\n",
       "  read_type                                         input_path  \\\n",
       "0        pe  /nfs/bioinformatics/pipeline/Illumina/241231_V...   \n",
       "1        pe  /nfs/bioinformatics/pipeline/Illumina/241231_V...   \n",
       "2        pe  /nfs/bioinformatics/pipeline/Illumina/241231_V...   \n",
       "3        pe  /nfs/bioinformatics/pipeline/Illumina/241231_V...   \n",
       "4        pe  /nfs/bioinformatics/pipeline/Illumina/241231_V...   \n",
       "\n",
       "                                         output_path flowcellID  \\\n",
       "0  /nfs/bioinformatics/pipeline/Illumina/241231_V...  AACN5FKHV   \n",
       "1  /nfs/bioinformatics/pipeline/Illumina/241231_V...  AACN5FKHV   \n",
       "2  /nfs/bioinformatics/pipeline/Illumina/241231_V...  AACN5FKHV   \n",
       "3  /nfs/bioinformatics/pipeline/Illumina/241231_V...  AACN5FKHV   \n",
       "4  /nfs/bioinformatics/pipeline/Illumina/241231_V...  AACN5FKHV   \n",
       "\n",
       "       sample_flowcell                                                 R1  \\\n",
       "0  lib105608_AACN5FKHV  ['/nfs/bioinformatics/pipeline/Illumina/241231...   \n",
       "1  lib105609_AACN5FKHV  ['/nfs/bioinformatics/pipeline/Illumina/241231...   \n",
       "2  lib105610_AACN5FKHV  ['/nfs/bioinformatics/pipeline/Illumina/241231...   \n",
       "3  lib105611_AACN5FKHV  ['/nfs/bioinformatics/pipeline/Illumina/241231...   \n",
       "4  lib105612_AACN5FKHV  ['/nfs/bioinformatics/pipeline/Illumina/241231...   \n",
       "\n",
       "                                                  R2       species     genome  \\\n",
       "0  ['/nfs/bioinformatics/pipeline/Illumina/241231...  Mus_musculus  GRCm38.91   \n",
       "1  ['/nfs/bioinformatics/pipeline/Illumina/241231...  Mus_musculus  GRCm38.91   \n",
       "2  ['/nfs/bioinformatics/pipeline/Illumina/241231...  Mus_musculus  GRCm38.91   \n",
       "3  ['/nfs/bioinformatics/pipeline/Illumina/241231...  Mus_musculus  GRCm38.91   \n",
       "4  ['/nfs/bioinformatics/pipeline/Illumina/241231...  Mus_musculus  GRCm38.91   \n",
       "\n",
       "   save_to_db                                       output_files  \\\n",
       "0        True  ['/nfs/bioinformatics/pipeline/Illumina/241231...   \n",
       "1        True  ['/nfs/bioinformatics/pipeline/Illumina/241231...   \n",
       "2        True  ['/nfs/bioinformatics/pipeline/Illumina/241231...   \n",
       "3        True  ['/nfs/bioinformatics/pipeline/Illumina/241231...   \n",
       "4        True  ['/nfs/bioinformatics/pipeline/Illumina/241231...   \n",
       "\n",
       "                                    genomics_samples        workflowbatchId  \\\n",
       "0  ['/nfs/bioinformatics/pipeline/Illumina/241231...  bri_P505-1_2025-01-02   \n",
       "1  ['/nfs/bioinformatics/pipeline/Illumina/241231...  bri_P505-1_2025-01-02   \n",
       "2  ['/nfs/bioinformatics/pipeline/Illumina/241231...  bri_P505-1_2025-01-02   \n",
       "3  ['/nfs/bioinformatics/pipeline/Illumina/241231...  bri_P505-1_2025-01-02   \n",
       "4  ['/nfs/bioinformatics/pipeline/Illumina/241231...  bri_P505-1_2025-01-02   \n",
       "\n",
       "                                           batchFile  \n",
       "0  /nfs/bioinformatics/pipeline/snakemakeLogs/202...  \n",
       "1  /nfs/bioinformatics/pipeline/snakemakeLogs/202...  \n",
       "2  /nfs/bioinformatics/pipeline/snakemakeLogs/202...  \n",
       "3  /nfs/bioinformatics/pipeline/snakemakeLogs/202...  \n",
       "4  /nfs/bioinformatics/pipeline/snakemakeLogs/202...  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_sheet.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_sheet['genome'] = 'Crassostrea_gigas_v1'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>project</th>\n",
       "      <th>flowcell</th>\n",
       "      <th>reference</th>\n",
       "      <th>assay_type</th>\n",
       "      <th>read_type</th>\n",
       "      <th>input_path</th>\n",
       "      <th>output_path</th>\n",
       "      <th>flowcellID</th>\n",
       "      <th>sample_flowcell</th>\n",
       "      <th>R1</th>\n",
       "      <th>R2</th>\n",
       "      <th>species</th>\n",
       "      <th>genome</th>\n",
       "      <th>save_to_db</th>\n",
       "      <th>output_files</th>\n",
       "      <th>genomics_samples</th>\n",
       "      <th>workflowbatchId</th>\n",
       "      <th>batchFile</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>lib105608</td>\n",
       "      <td>P505-1</td>\n",
       "      <td>241231_VH00126_375_AACN5FKHV</td>\n",
       "      <td>mouse</td>\n",
       "      <td>bulkRNAseq</td>\n",
       "      <td>pe</td>\n",
       "      <td>/nfs/bioinformatics/pipeline/Illumina/241231_V...</td>\n",
       "      <td>/nfs/bioinformatics/pipeline/Illumina/241231_V...</td>\n",
       "      <td>AACN5FKHV</td>\n",
       "      <td>lib105608_AACN5FKHV</td>\n",
       "      <td>['/nfs/bioinformatics/pipeline/Illumina/241231...</td>\n",
       "      <td>['/nfs/bioinformatics/pipeline/Illumina/241231...</td>\n",
       "      <td>Mus_musculus</td>\n",
       "      <td>Crassostrea_gigas_v1</td>\n",
       "      <td>True</td>\n",
       "      <td>['/nfs/bioinformatics/pipeline/Illumina/241231...</td>\n",
       "      <td>['/nfs/bioinformatics/pipeline/Illumina/241231...</td>\n",
       "      <td>bri_P505-1_2025-01-02</td>\n",
       "      <td>/nfs/bioinformatics/pipeline/snakemakeLogs/202...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>lib105609</td>\n",
       "      <td>P505-1</td>\n",
       "      <td>241231_VH00126_375_AACN5FKHV</td>\n",
       "      <td>mouse</td>\n",
       "      <td>bulkRNAseq</td>\n",
       "      <td>pe</td>\n",
       "      <td>/nfs/bioinformatics/pipeline/Illumina/241231_V...</td>\n",
       "      <td>/nfs/bioinformatics/pipeline/Illumina/241231_V...</td>\n",
       "      <td>AACN5FKHV</td>\n",
       "      <td>lib105609_AACN5FKHV</td>\n",
       "      <td>['/nfs/bioinformatics/pipeline/Illumina/241231...</td>\n",
       "      <td>['/nfs/bioinformatics/pipeline/Illumina/241231...</td>\n",
       "      <td>Mus_musculus</td>\n",
       "      <td>Crassostrea_gigas_v1</td>\n",
       "      <td>True</td>\n",
       "      <td>['/nfs/bioinformatics/pipeline/Illumina/241231...</td>\n",
       "      <td>['/nfs/bioinformatics/pipeline/Illumina/241231...</td>\n",
       "      <td>bri_P505-1_2025-01-02</td>\n",
       "      <td>/nfs/bioinformatics/pipeline/snakemakeLogs/202...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>lib105610</td>\n",
       "      <td>P505-1</td>\n",
       "      <td>241231_VH00126_375_AACN5FKHV</td>\n",
       "      <td>mouse</td>\n",
       "      <td>bulkRNAseq</td>\n",
       "      <td>pe</td>\n",
       "      <td>/nfs/bioinformatics/pipeline/Illumina/241231_V...</td>\n",
       "      <td>/nfs/bioinformatics/pipeline/Illumina/241231_V...</td>\n",
       "      <td>AACN5FKHV</td>\n",
       "      <td>lib105610_AACN5FKHV</td>\n",
       "      <td>['/nfs/bioinformatics/pipeline/Illumina/241231...</td>\n",
       "      <td>['/nfs/bioinformatics/pipeline/Illumina/241231...</td>\n",
       "      <td>Mus_musculus</td>\n",
       "      <td>Crassostrea_gigas_v1</td>\n",
       "      <td>True</td>\n",
       "      <td>['/nfs/bioinformatics/pipeline/Illumina/241231...</td>\n",
       "      <td>['/nfs/bioinformatics/pipeline/Illumina/241231...</td>\n",
       "      <td>bri_P505-1_2025-01-02</td>\n",
       "      <td>/nfs/bioinformatics/pipeline/snakemakeLogs/202...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>lib105611</td>\n",
       "      <td>P505-1</td>\n",
       "      <td>241231_VH00126_375_AACN5FKHV</td>\n",
       "      <td>mouse</td>\n",
       "      <td>bulkRNAseq</td>\n",
       "      <td>pe</td>\n",
       "      <td>/nfs/bioinformatics/pipeline/Illumina/241231_V...</td>\n",
       "      <td>/nfs/bioinformatics/pipeline/Illumina/241231_V...</td>\n",
       "      <td>AACN5FKHV</td>\n",
       "      <td>lib105611_AACN5FKHV</td>\n",
       "      <td>['/nfs/bioinformatics/pipeline/Illumina/241231...</td>\n",
       "      <td>['/nfs/bioinformatics/pipeline/Illumina/241231...</td>\n",
       "      <td>Mus_musculus</td>\n",
       "      <td>Crassostrea_gigas_v1</td>\n",
       "      <td>True</td>\n",
       "      <td>['/nfs/bioinformatics/pipeline/Illumina/241231...</td>\n",
       "      <td>['/nfs/bioinformatics/pipeline/Illumina/241231...</td>\n",
       "      <td>bri_P505-1_2025-01-02</td>\n",
       "      <td>/nfs/bioinformatics/pipeline/snakemakeLogs/202...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>lib105612</td>\n",
       "      <td>P505-1</td>\n",
       "      <td>241231_VH00126_375_AACN5FKHV</td>\n",
       "      <td>mouse</td>\n",
       "      <td>bulkRNAseq</td>\n",
       "      <td>pe</td>\n",
       "      <td>/nfs/bioinformatics/pipeline/Illumina/241231_V...</td>\n",
       "      <td>/nfs/bioinformatics/pipeline/Illumina/241231_V...</td>\n",
       "      <td>AACN5FKHV</td>\n",
       "      <td>lib105612_AACN5FKHV</td>\n",
       "      <td>['/nfs/bioinformatics/pipeline/Illumina/241231...</td>\n",
       "      <td>['/nfs/bioinformatics/pipeline/Illumina/241231...</td>\n",
       "      <td>Mus_musculus</td>\n",
       "      <td>Crassostrea_gigas_v1</td>\n",
       "      <td>True</td>\n",
       "      <td>['/nfs/bioinformatics/pipeline/Illumina/241231...</td>\n",
       "      <td>['/nfs/bioinformatics/pipeline/Illumina/241231...</td>\n",
       "      <td>bri_P505-1_2025-01-02</td>\n",
       "      <td>/nfs/bioinformatics/pipeline/snakemakeLogs/202...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      sample project                      flowcell reference  assay_type  \\\n",
       "0  lib105608  P505-1  241231_VH00126_375_AACN5FKHV     mouse  bulkRNAseq   \n",
       "1  lib105609  P505-1  241231_VH00126_375_AACN5FKHV     mouse  bulkRNAseq   \n",
       "2  lib105610  P505-1  241231_VH00126_375_AACN5FKHV     mouse  bulkRNAseq   \n",
       "3  lib105611  P505-1  241231_VH00126_375_AACN5FKHV     mouse  bulkRNAseq   \n",
       "4  lib105612  P505-1  241231_VH00126_375_AACN5FKHV     mouse  bulkRNAseq   \n",
       "\n",
       "  read_type                                         input_path  \\\n",
       "0        pe  /nfs/bioinformatics/pipeline/Illumina/241231_V...   \n",
       "1        pe  /nfs/bioinformatics/pipeline/Illumina/241231_V...   \n",
       "2        pe  /nfs/bioinformatics/pipeline/Illumina/241231_V...   \n",
       "3        pe  /nfs/bioinformatics/pipeline/Illumina/241231_V...   \n",
       "4        pe  /nfs/bioinformatics/pipeline/Illumina/241231_V...   \n",
       "\n",
       "                                         output_path flowcellID  \\\n",
       "0  /nfs/bioinformatics/pipeline/Illumina/241231_V...  AACN5FKHV   \n",
       "1  /nfs/bioinformatics/pipeline/Illumina/241231_V...  AACN5FKHV   \n",
       "2  /nfs/bioinformatics/pipeline/Illumina/241231_V...  AACN5FKHV   \n",
       "3  /nfs/bioinformatics/pipeline/Illumina/241231_V...  AACN5FKHV   \n",
       "4  /nfs/bioinformatics/pipeline/Illumina/241231_V...  AACN5FKHV   \n",
       "\n",
       "       sample_flowcell                                                 R1  \\\n",
       "0  lib105608_AACN5FKHV  ['/nfs/bioinformatics/pipeline/Illumina/241231...   \n",
       "1  lib105609_AACN5FKHV  ['/nfs/bioinformatics/pipeline/Illumina/241231...   \n",
       "2  lib105610_AACN5FKHV  ['/nfs/bioinformatics/pipeline/Illumina/241231...   \n",
       "3  lib105611_AACN5FKHV  ['/nfs/bioinformatics/pipeline/Illumina/241231...   \n",
       "4  lib105612_AACN5FKHV  ['/nfs/bioinformatics/pipeline/Illumina/241231...   \n",
       "\n",
       "                                                  R2       species  \\\n",
       "0  ['/nfs/bioinformatics/pipeline/Illumina/241231...  Mus_musculus   \n",
       "1  ['/nfs/bioinformatics/pipeline/Illumina/241231...  Mus_musculus   \n",
       "2  ['/nfs/bioinformatics/pipeline/Illumina/241231...  Mus_musculus   \n",
       "3  ['/nfs/bioinformatics/pipeline/Illumina/241231...  Mus_musculus   \n",
       "4  ['/nfs/bioinformatics/pipeline/Illumina/241231...  Mus_musculus   \n",
       "\n",
       "                 genome  save_to_db  \\\n",
       "0  Crassostrea_gigas_v1        True   \n",
       "1  Crassostrea_gigas_v1        True   \n",
       "2  Crassostrea_gigas_v1        True   \n",
       "3  Crassostrea_gigas_v1        True   \n",
       "4  Crassostrea_gigas_v1        True   \n",
       "\n",
       "                                        output_files  \\\n",
       "0  ['/nfs/bioinformatics/pipeline/Illumina/241231...   \n",
       "1  ['/nfs/bioinformatics/pipeline/Illumina/241231...   \n",
       "2  ['/nfs/bioinformatics/pipeline/Illumina/241231...   \n",
       "3  ['/nfs/bioinformatics/pipeline/Illumina/241231...   \n",
       "4  ['/nfs/bioinformatics/pipeline/Illumina/241231...   \n",
       "\n",
       "                                    genomics_samples        workflowbatchId  \\\n",
       "0  ['/nfs/bioinformatics/pipeline/Illumina/241231...  bri_P505-1_2025-01-02   \n",
       "1  ['/nfs/bioinformatics/pipeline/Illumina/241231...  bri_P505-1_2025-01-02   \n",
       "2  ['/nfs/bioinformatics/pipeline/Illumina/241231...  bri_P505-1_2025-01-02   \n",
       "3  ['/nfs/bioinformatics/pipeline/Illumina/241231...  bri_P505-1_2025-01-02   \n",
       "4  ['/nfs/bioinformatics/pipeline/Illumina/241231...  bri_P505-1_2025-01-02   \n",
       "\n",
       "                                           batchFile  \n",
       "0  /nfs/bioinformatics/pipeline/snakemakeLogs/202...  \n",
       "1  /nfs/bioinformatics/pipeline/snakemakeLogs/202...  \n",
       "2  /nfs/bioinformatics/pipeline/snakemakeLogs/202...  \n",
       "3  /nfs/bioinformatics/pipeline/snakemakeLogs/202...  \n",
       "4  /nfs/bioinformatics/pipeline/snakemakeLogs/202...  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_sheet.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "def build_reference_dict(annotation_dir):\n",
    " \n",
    "    # Initialize the reference dictionary\n",
    "    reference_dict = {}\n",
    "\n",
    "    # List all genome directories (excluding irrelevant entries)\n",
    "    genome_dirs = [d for d in os.listdir(annotation_dir) if os.path.isdir(os.path.join(annotation_dir, d))]\n",
    "\n",
    "    # Loop through each genome directory\n",
    "    for genome in genome_dirs:\n",
    "        genome_dir = os.path.join(annotation_dir, genome)\n",
    "\n",
    "        # Find files based on their patterns\n",
    "        gtf_ref = glob.glob(os.path.join(genome_dir, \"*.gtf\"))\n",
    "        ref_fa = glob.glob(os.path.join(genome_dir, \"*.fa\"))\n",
    "        refflat_ref = glob.glob(os.path.join(genome_dir, \"*refflat.txt\"))\n",
    "        ribo_ref = glob.glob(os.path.join(genome_dir, \"*ribosomalIntervalsWheader.txt\"))\n",
    "        chromap_index = glob.glob(os.path.join(genome_dir, \"*chromap\"))\n",
    "        snps_bed = glob.glob(os.path.join(genome_dir, \"*NGSCheckMate_andInterestingSNP.bed\"))\n",
    "\n",
    "        # Construct the dictionary for this genome by type\n",
    "        reference_dict[genome] = {\n",
    "            \"star_align\": os.path.join(genome_dir, \"star\") if os.path.exists(os.path.join(genome_dir, \"star\")) else None,\n",
    "            \"htseq_count\": gtf_ref[0] if gtf_ref else None,\n",
    "            \"picard_metrics\": {\n",
    "                \"fa\": ref_fa[0] if ref_fa else None,\n",
    "                \"flat\": refflat_ref[0] if refflat_ref else None,\n",
    "                \"ribosomal_intervals\": ribo_ref[0] if ribo_ref else None,\n",
    "            },\n",
    "            \"snps\": {\n",
    "                \"fa\": ref_fa[0] if ref_fa else None,\n",
    "                \"snps_bed\": snps_bed[0] if snps_bed else None,\n",
    "            } if \"GRCh\" in genome or \"hg\" in genome else None,  # SNPs only for human genomes\n",
    "            \"chromap\": {\n",
    "                \"fa\": ref_fa[0] if ref_fa else None,\n",
    "                \"index\": chromap_index[0] if chromap_index else None,\n",
    "            },\n",
    "            \"macs\": \"hs\" if \"GRCh\" in genome or \"hg\" in genome else \"mm\" if \"GRCm\" in genome or \"mm\" in genome else None,\n",
    "            \"bigwigs\": \"2913022398\" if \"GRCh\" in genome or \"hg\" in genome else \"2652783500\" if \"GRCm\" in genome or \"mm\" in genome else None,\n",
    "        }\n",
    "\n",
    "    return reference_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "reference_dict = build_reference_dict(ANNOTATIONS_DIR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "def genome_picker_v3(type, sample):\n",
    "    # Get genome for the current sample\n",
    "    try:\n",
    "        genome = sample_sheet[\"genome\"].loc[sample_sheet[\"sample_flowcell\"] == sample].item()\n",
    "    except ValueError:\n",
    "        genome = sample_sheet[\"genome\"].loc[sample_sheet[\"sample\"] == sample].item()\n",
    "\n",
    "    reference = reference_dict[genome][type]\n",
    "\n",
    "    return reference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'containers': {'star_align': None,\n",
       "  'htseq_count': None,\n",
       "  'picard_metrics': {'fa': None, 'flat': None, 'ribosomal_intervals': None},\n",
       "  'snps': None,\n",
       "  'chromap': {'fa': None, 'index': None},\n",
       "  'macs': None,\n",
       "  'bigwigs': None},\n",
       " 'ICACviruses': {'star_align': '/Volumes/bioinformatics/pipeline/annotation/pipeline/ICACviruses/star',\n",
       "  'htseq_count': None,\n",
       "  'picard_metrics': {'fa': '/Volumes/bioinformatics/pipeline/annotation/pipeline/ICACviruses/241025_ICACViruses.fa',\n",
       "   'flat': None,\n",
       "   'ribosomal_intervals': None},\n",
       "  'snps': None,\n",
       "  'chromap': {'fa': '/Volumes/bioinformatics/pipeline/annotation/pipeline/ICACviruses/241025_ICACViruses.fa',\n",
       "   'index': None},\n",
       "  'macs': None,\n",
       "  'bigwigs': None},\n",
       " '10x': {'star_align': None,\n",
       "  'htseq_count': None,\n",
       "  'picard_metrics': {'fa': None, 'flat': None, 'ribosomal_intervals': None},\n",
       "  'snps': None,\n",
       "  'chromap': {'fa': None, 'index': None},\n",
       "  'macs': None,\n",
       "  'bigwigs': None},\n",
       " 'Crassostrea_virginica_3.0': {'star_align': '/Volumes/bioinformatics/pipeline/annotation/pipeline/Crassostrea_virginica_3.0/star',\n",
       "  'htseq_count': '/Volumes/bioinformatics/pipeline/annotation/pipeline/Crassostrea_virginica_3.0/Crassostrea_virginica_3.0.gtf',\n",
       "  'picard_metrics': {'fa': None,\n",
       "   'flat': None,\n",
       "   'ribosomal_intervals': '/Volumes/bioinformatics/pipeline/annotation/pipeline/Crassostrea_virginica_3.0/Crassostrea_virginica_3.0.ribosomalIntervalsWheader.txt'},\n",
       "  'snps': None,\n",
       "  'chromap': {'fa': None, 'index': None},\n",
       "  'macs': None,\n",
       "  'bigwigs': None},\n",
       " 'GRCm38.91_tdTomato_ZsGreen': {'star_align': '/Volumes/bioinformatics/pipeline/annotation/pipeline/GRCm38.91_tdTomato_ZsGreen/star',\n",
       "  'htseq_count': '/Volumes/bioinformatics/pipeline/annotation/pipeline/GRCm38.91_tdTomato_ZsGreen/GRCm38.91_tdTomato_ZsGreen.gtf',\n",
       "  'picard_metrics': {'fa': '/Volumes/bioinformatics/pipeline/annotation/pipeline/GRCm38.91_tdTomato_ZsGreen/GRCm38.91_tdTomato_ZsGreen.fa',\n",
       "   'flat': '/Volumes/bioinformatics/pipeline/annotation/pipeline/GRCm38.91_tdTomato_ZsGreen/GRCm38.91_tdTomato_ZsGreen.refflat.txt',\n",
       "   'ribosomal_intervals': '/Volumes/bioinformatics/pipeline/annotation/pipeline/GRCm38.91_tdTomato_ZsGreen/GRCm38.91_tdTomato_ZsGreen.ribosomalIntervalsWheader.txt'},\n",
       "  'snps': None,\n",
       "  'chromap': {'fa': '/Volumes/bioinformatics/pipeline/annotation/pipeline/GRCm38.91_tdTomato_ZsGreen/GRCm38.91_tdTomato_ZsGreen.fa',\n",
       "   'index': None},\n",
       "  'macs': 'mm',\n",
       "  'bigwigs': '2652783500'},\n",
       " 'GRCm38.91_tdTomato': {'star_align': '/Volumes/bioinformatics/pipeline/annotation/pipeline/GRCm38.91_tdTomato/star',\n",
       "  'htseq_count': '/Volumes/bioinformatics/pipeline/annotation/pipeline/GRCm38.91_tdTomato/GRCm38.91_tdTomato.gtf',\n",
       "  'picard_metrics': {'fa': '/Volumes/bioinformatics/pipeline/annotation/pipeline/GRCm38.91_tdTomato/GRCm38.91_tdTomato.fa',\n",
       "   'flat': '/Volumes/bioinformatics/pipeline/annotation/pipeline/GRCm38.91_tdTomato/GRCm38.91_tdTomato.refflat.txt',\n",
       "   'ribosomal_intervals': '/Volumes/bioinformatics/pipeline/annotation/pipeline/GRCm38.91_tdTomato/GRCm38.91_tdTomato.ribosomalIntervalsWheader.txt'},\n",
       "  'snps': None,\n",
       "  'chromap': {'fa': '/Volumes/bioinformatics/pipeline/annotation/pipeline/GRCm38.91_tdTomato/GRCm38.91_tdTomato.fa',\n",
       "   'index': None},\n",
       "  'macs': 'mm',\n",
       "  'bigwigs': '2652783500'},\n",
       " 'GRCm38.91': {'star_align': '/Volumes/bioinformatics/pipeline/annotation/pipeline/GRCm38.91/star',\n",
       "  'htseq_count': '/Volumes/bioinformatics/pipeline/annotation/pipeline/GRCm38.91/Mus_musculus.GRCm38.91.gtf',\n",
       "  'picard_metrics': {'fa': '/Volumes/bioinformatics/pipeline/annotation/pipeline/GRCm38.91/Mus_musculus.GRCm38.dna.primary_assembly.fa',\n",
       "   'flat': '/Volumes/bioinformatics/pipeline/annotation/pipeline/GRCm38.91/Mus_musculus.GRCm38.91.gtf.refflat.txt',\n",
       "   'ribosomal_intervals': '/Volumes/bioinformatics/pipeline/annotation/pipeline/GRCm38.91/Mus_musculus.GRCm38.91.ribosomalIntervalsWheader.txt'},\n",
       "  'snps': None,\n",
       "  'chromap': {'fa': '/Volumes/bioinformatics/pipeline/annotation/pipeline/GRCm38.91/Mus_musculus.GRCm38.dna.primary_assembly.fa',\n",
       "   'index': '/Volumes/bioinformatics/pipeline/annotation/pipeline/GRCm38.91/Mus_musculus.GRCm38.dna.primary_assembly.chromap'},\n",
       "  'macs': 'mm',\n",
       "  'bigwigs': '2652783500'},\n",
       " 'GRCh38.91': {'star_align': '/Volumes/bioinformatics/pipeline/annotation/pipeline/GRCh38.91/star',\n",
       "  'htseq_count': '/Volumes/bioinformatics/pipeline/annotation/pipeline/GRCh38.91/Homo_sapiens.GRCh38.91.gtf',\n",
       "  'picard_metrics': {'fa': '/Volumes/bioinformatics/pipeline/annotation/pipeline/GRCh38.91/Homo_sapiens.GRCh38.dna.primary_assembly.fa',\n",
       "   'flat': '/Volumes/bioinformatics/pipeline/annotation/pipeline/GRCh38.91/GRCh38.91.refflat.txt',\n",
       "   'ribosomal_intervals': '/Volumes/bioinformatics/pipeline/annotation/pipeline/GRCh38.91/GRCh38.91.ribosomalIntervalsWheader.txt'},\n",
       "  'snps': {'fa': '/Volumes/bioinformatics/pipeline/annotation/pipeline/GRCh38.91/Homo_sapiens.GRCh38.dna.primary_assembly.fa',\n",
       "   'snps_bed': '/Volumes/bioinformatics/pipeline/annotation/pipeline/GRCh38.91/GRCh38_NGSCheckMate_andInterestingSNP.bed'},\n",
       "  'chromap': {'fa': '/Volumes/bioinformatics/pipeline/annotation/pipeline/GRCh38.91/Homo_sapiens.GRCh38.dna.primary_assembly.fa',\n",
       "   'index': '/Volumes/bioinformatics/pipeline/annotation/pipeline/GRCh38.91/Homo_sapiens.GRCh38.dna.primary_assembly.chromap'},\n",
       "  'macs': 'hs',\n",
       "  'bigwigs': '2913022398'},\n",
       " 'TyGRCh38': {'star_align': None,\n",
       "  'htseq_count': None,\n",
       "  'picard_metrics': {'fa': '/Volumes/bioinformatics/pipeline/annotation/pipeline/TyGRCh38/TyGRCh38.fa',\n",
       "   'flat': None,\n",
       "   'ribosomal_intervals': None},\n",
       "  'snps': {'fa': '/Volumes/bioinformatics/pipeline/annotation/pipeline/TyGRCh38/TyGRCh38.fa',\n",
       "   'snps_bed': None},\n",
       "  'chromap': {'fa': '/Volumes/bioinformatics/pipeline/annotation/pipeline/TyGRCh38/TyGRCh38.fa',\n",
       "   'index': '/Volumes/bioinformatics/pipeline/annotation/pipeline/TyGRCh38/TyGRCh38.chromap'},\n",
       "  'macs': 'hs',\n",
       "  'bigwigs': '2913022398'},\n",
       " 'Crassostrea_gigas_v1': {'star_align': '/Volumes/bioinformatics/pipeline/annotation/pipeline/Crassostrea_gigas_v1/star',\n",
       "  'htseq_count': '/Volumes/bioinformatics/pipeline/annotation/pipeline/Crassostrea_gigas_v1/Crassostrea_gigas_v1.gtf',\n",
       "  'picard_metrics': {'fa': '/Volumes/bioinformatics/pipeline/annotation/pipeline/Crassostrea_gigas_v1/Crassostrea_gigas_v1.fa',\n",
       "   'flat': '/Volumes/bioinformatics/pipeline/annotation/pipeline/Crassostrea_gigas_v1/Crassostrea_gigas_v1.refflat.txt',\n",
       "   'ribosomal_intervals': '/Volumes/bioinformatics/pipeline/annotation/pipeline/Crassostrea_gigas_v1/Crassostrea_gigas_v1.ribosomalIntervalsWheader.txt'},\n",
       "  'snps': None,\n",
       "  'chromap': {'fa': '/Volumes/bioinformatics/pipeline/annotation/pipeline/Crassostrea_gigas_v1/Crassostrea_gigas_v1.fa',\n",
       "   'index': None},\n",
       "  'macs': None,\n",
       "  'bigwigs': None},\n",
       " 'GRCh37': {'star_align': '/Volumes/bioinformatics/pipeline/annotation/pipeline/GRCh37/star',\n",
       "  'htseq_count': '/Volumes/bioinformatics/pipeline/annotation/pipeline/GRCh37/Homo_sapiens.GRCh37.87.gtf',\n",
       "  'picard_metrics': {'fa': '/Volumes/bioinformatics/pipeline/annotation/pipeline/GRCh37/Homo_sapiens.GRCh37.dna.primary_assembly.fa',\n",
       "   'flat': '/Volumes/bioinformatics/pipeline/annotation/pipeline/GRCh37/Homo_sapiens.GRCh37.87.refflat.txt',\n",
       "   'ribosomal_intervals': None},\n",
       "  'snps': {'fa': '/Volumes/bioinformatics/pipeline/annotation/pipeline/GRCh37/Homo_sapiens.GRCh37.dna.primary_assembly.fa',\n",
       "   'snps_bed': None},\n",
       "  'chromap': {'fa': '/Volumes/bioinformatics/pipeline/annotation/pipeline/GRCh37/Homo_sapiens.GRCh37.dna.primary_assembly.fa',\n",
       "   'index': None},\n",
       "  'macs': 'hs',\n",
       "  'bigwigs': '2913022398'},\n",
       " 'GRCm38.91_SS78_SS95': {'star_align': '/Volumes/bioinformatics/pipeline/annotation/pipeline/GRCm38.91_SS78_SS95/star',\n",
       "  'htseq_count': '/Volumes/bioinformatics/pipeline/annotation/pipeline/GRCm38.91_SS78_SS95/GRCm38.91_SS78_SS95.gtf',\n",
       "  'picard_metrics': {'fa': '/Volumes/bioinformatics/pipeline/annotation/pipeline/GRCm38.91_SS78_SS95/SS78.fa',\n",
       "   'flat': '/Volumes/bioinformatics/pipeline/annotation/pipeline/GRCm38.91_SS78_SS95/GRCm38.91_SS78_SS95.gtf.refflat.txt',\n",
       "   'ribosomal_intervals': '/Volumes/bioinformatics/pipeline/annotation/pipeline/GRCm38.91_SS78_SS95/GRCm38.91_SS78_SS95.ribosomalIntervalsWheader.txt'},\n",
       "  'snps': None,\n",
       "  'chromap': {'fa': '/Volumes/bioinformatics/pipeline/annotation/pipeline/GRCm38.91_SS78_SS95/SS78.fa',\n",
       "   'index': None},\n",
       "  'macs': 'mm',\n",
       "  'bigwigs': '2652783500'},\n",
       " 'GRCh37_ucsc': {'star_align': '/Volumes/bioinformatics/pipeline/annotation/pipeline/GRCh37_ucsc/star',\n",
       "  'htseq_count': '/Volumes/bioinformatics/pipeline/annotation/pipeline/GRCh37_ucsc/hg19.refGene.gtf',\n",
       "  'picard_metrics': {'fa': '/Volumes/bioinformatics/pipeline/annotation/pipeline/GRCh37_ucsc/hg19.fa',\n",
       "   'flat': None,\n",
       "   'ribosomal_intervals': None},\n",
       "  'snps': {'fa': '/Volumes/bioinformatics/pipeline/annotation/pipeline/GRCh37_ucsc/hg19.fa',\n",
       "   'snps_bed': None},\n",
       "  'chromap': {'fa': '/Volumes/bioinformatics/pipeline/annotation/pipeline/GRCh37_ucsc/hg19.fa',\n",
       "   'index': None},\n",
       "  'macs': 'hs',\n",
       "  'bigwigs': '2913022398'}}"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reference_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/Volumes/bioinformatics/pipeline/annotation/pipeline/Crassostrea_gigas_v1/star'"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genome_picker_v3(\"star_align\",\"lib105609\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['containers',\n",
       " 'ICACviruses',\n",
       " '10x',\n",
       " 'Crassostrea_virginica_3.0',\n",
       " 'GRCm38.91_tdTomato_ZsGreen',\n",
       " 'GRCm38.91_tdTomato',\n",
       " 'GRCm38.91',\n",
       " 'GRCh38.91',\n",
       " 'TyGRCh38',\n",
       " 'Crassostrea_gigas_v1',\n",
       " 'GRCh37',\n",
       " 'GRCm38.91_SS78_SS95',\n",
       " 'GRCh37_ucsc']"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[d for d in os.listdir(ANNOTATIONS_DIR) if not d.startswith(\".\") and os.path.isdir(os.path.join(ANNOTATIONS_DIR, d))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['/Volumes/bioinformatics/pipeline/annotation/pipeline/GRCm38.91/Mus_musculus.GRCm38.91.gtf']"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "glob.glob(os.path.join(os.path.join(ANNOTATIONS_DIR, \"GRCm38.91\"), \"*.gtf\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "reference_dict = build_reference_dict(ANNOTATIONS_DIR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Volumes/bioinformatics/pipeline/annotation/pipeline/Crassostrea_gigas_v1/Crassostrea_gigas_v1.gtf\n"
     ]
    }
   ],
   "source": [
    "print(reference_dict[\"Crassostrea_gigas_v1\"][\"gtf\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def select_genome(row, defaultGenomes, default10xReferences):\n",
    "\n",
    "    if row[\"reference\"] == \"human\":\n",
    "        species = \"Homo_sapiens\"\n",
    "    elif row[\"reference\"] == \"mouse\":\n",
    "        species = \"Mus_musculus\"\n",
    "    else:\n",
    "        species = \"\"\n",
    "        genome = row[\"reference\"]\n",
    "\n",
    "        # if neither human nor mouse, return the reference that was given\n",
    "        return species, genome\n",
    "    \n",
    "    if row[\"assay_type\"] != \"10x\":\n",
    "        genome = defaultGenomes[row[\"reference\"]]\n",
    "    else:\n",
    "        genome = default10xReferences[row[\"reference\"]][\"build\"]\n",
    "        \n",
    "\n",
    "    return species, genome\n",
    "\n",
    "def genome_picker(type, sample):\n",
    "    # if no \n",
    "\n",
    "    # selecting genome build and species for the current library being run on pipeline\n",
    "    try:\n",
    "        species = sample_sheet[\"species\"].loc[sample_sheet[\"sample_flowcell\"] == sample].item() \n",
    "        genome = sample_sheet[\"genome\"].loc[sample_sheet[\"sample_flowcell\"] == sample].item()\n",
    "    except ValueError:\n",
    "        species = sample_sheet[\"species\"].loc[sample_sheet[\"sample\"] == sample].item()\n",
    "        genome = sample_sheet[\"genome\"].loc[sample_sheet[\"sample\"] == sample].item()\n",
    "\n",
    "    if species == \"\":\n",
    "        gtf_ref =  \".\".join([genome, \"gtf\"])\n",
    "        ref_fa = \".\".join([genome, \"fa\"]) # instead of indexing genome, we should probably do a split or regexp\n",
    "        refflat_ref = \".\".join([genome, \"refflat.txt\"])\n",
    "        ribo_ref = \".\".join([genome, \"ribosomalIntervalsWheader.txt\"])\n",
    "        chromap_index = \".\".join([genome, \"chromap\"])\n",
    "    else:\n",
    "        gtf_ref =  \".\".join([species, genome, \"gtf\"])\n",
    "        ref_fa = \".\".join([species, genome[0:6], \"dna.primary_assembly.fa\"]) # instead of indexing genome, we should probably do a split or regexp\n",
    "        refflat_ref = \".\".join([species, genome, \"refflat.txt\"])\n",
    "        ribo_ref = \".\".join([species, genome, \"ribosomalIntervalsWheader.txt\"])\n",
    "        chromap_index = \".\".join([species, genome[0:6], \"dna.primary_assembly.chromap\"])\n",
    "    \n",
    "    if type == \"star_align\":\n",
    "        reference = os.path.join(ANNOTATIONS_DIR, genome, \"star\") \n",
    "\n",
    "    elif type == \"htseq_count\":\n",
    "        reference = os.path.join(ANNOTATIONS_DIR, genome, gtf_ref)\n",
    "\n",
    "    elif type == \"picard_metrics\":\n",
    "        reference = { \"fa\":                 os.path.join(ANNOTATIONS_DIR, genome, ref_fa), \n",
    "                    \"flat\" :                os.path.join(ANNOTATIONS_DIR, genome, refflat_ref),\n",
    "                    \"ribosomal_intervals\" : os.path.join(ANNOTATIONS_DIR, genome, ribo_ref)}\n",
    "        \n",
    "    elif type == \"snps\":\n",
    "        reference = { \"fa\":     os.path.join(ANNOTATIONS_DIR, genome, ref_fa),\n",
    "                    \"snps_bed\": os.path.join(ANNOTATIONS_DIR, genome, \"GRCh38_NGSCheckMate_andInterestingSNP.bed\") } \n",
    "        \n",
    "    elif type == \"chromap\":\n",
    "        reference = { \"fa\":   os.path.join(ANNOTATIONS_DIR, genome, ref_fa), \n",
    "                    \"index\" : os.path.join(ANNOTATIONS_DIR, genome, chromap_index)}\n",
    "        \n",
    "    elif type == \"macs\":\n",
    "        if species == \"Homo_sapiens\":\n",
    "            reference = \"hs\"\n",
    "        elif species == \"Mus_musculus\":\n",
    "            reference = \"mm\"\n",
    "        else: \n",
    "            raise Exception(\"Support for custom genome sizes not implemented yet.\")\n",
    "        \n",
    "    elif type == \"bigwigs\":\n",
    "        if species == \"Homo_sapiens\":\n",
    "            reference = \"2913022398\"\n",
    "        elif species == \"Mus_musculus\":\n",
    "            reference = \"2652783500\"\n",
    "        else: \n",
    "            raise Exception(\"Support for custom genome sizes not implemneted yet.,\")\n",
    "\n",
    "    return (reference)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "def genome_picker_v2(type, sample):\n",
    "    # Get genome for the current sample\n",
    "    try:\n",
    "        genome = sample_sheet[\"genome\"].loc[sample_sheet[\"sample_flowcell\"] == sample].item()\n",
    "    except ValueError:\n",
    "        genome = sample_sheet[\"genome\"].loc[sample_sheet[\"sample\"] == sample].item()\n",
    "\n",
    "    # Define the base directory for annotations\n",
    "    genome_dir = os.path.join(ANNOTATIONS_DIR, genome)\n",
    "    \n",
    "    if not os.path.exists(genome_dir):\n",
    "        raise Exception(f\"Genome folder for {genome} does not exist in {ANNOTATIONS_DIR}\")\n",
    "\n",
    "    # Helper function to detect files\n",
    "    def find_file(pattern):\n",
    "        files = glob.glob(os.path.join(genome_dir, pattern))\n",
    "        if not files:\n",
    "            #logger.debug(f\"File matching pattern '{pattern}' not found in {genome_dir}\")\n",
    "            return None\n",
    "        return files[0]  # Return the first match\n",
    "\n",
    "    # Locate required files\n",
    "    gtf_ref = find_file(\"*.gtf\")\n",
    "    ref_fa = find_file(\"*.fa\")\n",
    "    refflat_ref = find_file(\"*refflat.txt\")\n",
    "    ribo_ref = find_file(\"*ribosomalIntervalsWheader.txt\")\n",
    "    chromap_index = find_file(\"*chromap\")\n",
    "\n",
    "    # Match type to the appropriate reference\n",
    "    if type == \"star_align\":\n",
    "        reference = os.path.join(genome_dir, \"star\")\n",
    "\n",
    "    elif type == \"htseq_count\":\n",
    "        reference = gtf_ref\n",
    "\n",
    "    elif type == \"picard_metrics\":\n",
    "        reference = {\n",
    "            \"fa\": ref_fa,\n",
    "            \"flat\": refflat_ref,\n",
    "            \"ribosomal_intervals\": ribo_ref,\n",
    "        }\n",
    "\n",
    "    elif type == \"snps\":\n",
    "        try:\n",
    "        # SNPs can only be done for human genomes\n",
    "            if \"GRCh\" not in genome and \"hg\" not in genome:\n",
    "                raise Exception(\"SNPs are currently supported only for human genomes (e.g., GRCh38 or hg19).\")\n",
    "\n",
    "            snps_bed = find_file(\"*NGSCheckMate_andInterestingSNP.bed\")\n",
    "            reference = {\n",
    "            \"fa\": ref_fa,\n",
    "            \"snps_bed\": snps_bed,\n",
    "            }\n",
    "        except Exception as e:\n",
    "            raise Exception(f\"Error in locating SNP references for genome '{genome}': {str(e)}\")\n",
    "\n",
    "\n",
    "\n",
    "    elif type == \"chromap\":\n",
    "        reference = {\n",
    "            \"fa\": ref_fa,\n",
    "            \"index\": chromap_index,\n",
    "        }\n",
    "\n",
    "    elif type == \"macs\":\n",
    "        if \"GRCh\" in genome or \"hg\" in genome:\n",
    "            reference = \"hs\"\n",
    "        elif \"GRCm\" in genome or \"mm\" in genome:\n",
    "            reference = \"mm\"\n",
    "        else:\n",
    "            raise Exception(\"Support for custom genome sizes not implemented yet.\")\n",
    "\n",
    "    elif type == \"bigwigs\":\n",
    "        if \"GRCh\" in genome or \"hg\" in genome:\n",
    "            reference = \"2913022398\"\n",
    "        elif \"GRCm\" in genome or \"mm\" in genome:\n",
    "            reference = \"2652783500\"\n",
    "        else:\n",
    "            raise Exception(\"Support for custom genome sizes not implemented yet.\")\n",
    "\n",
    "    return reference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/Volumes/bioinformatics/pipeline/annotation/pipeline/GRCm38.91/star'"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genome_picker(type=\"star_align\",sample=\"lib105608\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'fa': '/Volumes/bioinformatics/pipeline/annotation/pipeline/Crassostrea_gigas_v1/Crassostrea_gigas_v1.fa',\n",
       " 'flat': '/Volumes/bioinformatics/pipeline/annotation/pipeline/Crassostrea_gigas_v1/Crassostrea_gigas_v1.refflat.txt',\n",
       " 'ribosomal_intervals': '/Volumes/bioinformatics/pipeline/annotation/pipeline/Crassostrea_gigas_v1/Crassostrea_gigas_v1.ribosomalIntervalsWheader.txt'}"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genome_picker_v2(type=\"picard_metrics\",sample=\"lib105608\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    Mus_musculus\n",
       "1    Mus_musculus\n",
       "2    Mus_musculus\n",
       "3    Mus_musculus\n",
       "4    Mus_musculus\n",
       "5    Mus_musculus\n",
       "6    Mus_musculus\n",
       "7    Mus_musculus\n",
       "8    Mus_musculus\n",
       "Name: species, dtype: object"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "species = sample_sheet[\"species\"].loc[sample_sheet[\"sample_flowcell\"] == \"lib105608\"].item() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Mus_musculus'"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_sheet[\"species\"].loc[sample_sheet[\"sample\"]== \"lib105608\"].item() "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
